Realm/ObjectStore/object_store.cpp

```diff
@@ -0,0 +1,418 @@
+            if (current_prop.is_primary) {
+                validation_errors.push_back("Property '" + current_prop.name + "' is no longer a primary key.");
+            }
+            else {
+                validation_errors.push_back("Property '" + current_prop.name + "' has been made a primary key.");
+            }
+        }
+
+        // create new property with aligned column
+        target_prop->table_column = current_prop.table_column;
+    }
+
+    // check for new missing properties
+    for (auto& target_prop : target_schema.properties) {
+        if (!table_schema.property_for_name(target_prop.name)) {
+            validation_errors.push_back("Property '" + target_prop.name + "' has been added to latest object model.");
+        }
+    }
+
+    return validation_errors;
+}
+
+void ObjectStore::update_column_mapping(Group *group, ObjectSchema &target_schema) {
+    ObjectSchema table_schema(group, target_schema.name);
+    for (auto& target_prop : target_schema.properties) {
+        auto table_prop = table_schema.property_for_name(target_prop.name);
+        REALM_ASSERT_DEBUG(table_prop);
+
+        target_prop.table_column = table_prop->table_column;
+    }
+}
+
+static inline bool property_has_changed(Property &p1, Property &p2) {
+    return p1.type != p2.type || p1.name != p2.name || p1.object_type != p2.object_type;
+}
+
+// set references to tables on targetSchema and create/update any missing or out-of-date tables
+// if update existing is true, updates existing tables, otherwise validates existing tables
+// NOTE: must be called from within write transaction
+bool ObjectStore::create_tables(realm::Group *group, ObjectStore::Schema &target_schema, bool update_existing) {
+    bool changed = false;
+
+    // first pass to create missing tables
+    vector<ObjectSchema *> to_update;
+    set<string> target_type_names;
+    for (auto& object_schema : target_schema) {
+        bool created = false;
+        ObjectStore::table_for_object_type_create_if_needed(group, object_schema.name, created);
+
+        // we will modify tables for any new objectSchema (table was created) or for all if update_existing is true
+        if (update_existing || created) {
+            to_update.push_back(&object_schema);
+            changed = true;
+        }
+
+        // keep track of names to figure out what tables need deletion
+        target_type_names.insert(object_schema.name);
+    }
+
+    // second pass adds/removes columns for out of date tables
+    for (auto& target_object_schema : to_update) {
+        TableRef table = table_for_object_type(group, target_object_schema->name);
+        ObjectSchema current_schema(group, target_object_schema->name, table.get());
+        vector<Property> &target_props = target_object_schema->properties;
+
+        // add missing columns
+        for (auto& target_prop : target_props) {
+            auto current_prop = current_schema.property_for_name(target_prop.name);
```

Comment:
A column name must be less than 64 characters (I believe core will truncate). I think we should validate `target_prop.name`.